src/notifications/model.rs

```diff
@@ -183,13 +183,194 @@ impl Notifications {
+            NotificationsError::DatabaseError { inner } => inner.into(),
+        }
+    }
+}
+
+impl Notifications {
+    pub async fn set_notification_state<'a, E>(
+        payload: SetNotificationStatePayload,
+        executor: E,
+        pool: &MySqlPool,
+    ) -> Result<Notifications, SetNotificationStateError>
+    where
+        E: Executor<'a>,
+    {
+        let mut transaction = executor.begin().await?;
+
+        for id in payload.ids.into_iter() {
+            let seen = !payload.unread;
+            let notification = sqlx::query!(
+                r#"
+                    SELECT id, seen
+                        FROM notification
+                        WHERE id = ?
+                "#,
+                id
+            )
+            .fetch_one(&mut transaction)
+            .await?;
+
+            if (notification.seen != 0) == seen {
+                continue;
+            }
+
+            sqlx::query!(
+                r#"
+                    UPDATE notification
+                        SET seen = ?
+                        WHERE id = ?
+                "#,
+                seen,
+                id
+            )
+            .execute(&mut transaction)
+            .await?;
+        }
+
+        transaction.commit().await?;
+
+        Ok(Notifications::fetch_via_transaction(payload.user_id, pool).await?)
+    }
+}
+
 #[cfg(test)]
 mod tests {
-    use super::Notifications;
+    use super::{Notifications, SetNotificationStatePayload};
     use crate::create_database_pool;
     use crate::event::Event;
     use crate::subscriptions::Subscriptions;
 
+    #[actix_rt::test]
+    async fn set_notification_state_no_id() {
+        let pool = create_database_pool().await.unwrap();
+        let mut transaction = pool.begin().await.unwrap();
+
+        Notifications::set_notification_state(
+            SetNotificationStatePayload {
+                ids: vec![],
+                user_id: 1,
+                unread: true,
+            },
+            &mut transaction,
+            &pool,
+        )
+        .await
+        .unwrap();
+    }
+
+    #[actix_rt::test]
+    async fn set_notification_state_single_id() {
+        let pool = create_database_pool().await.unwrap();
+        let mut transaction = pool.begin().await.unwrap();
+
+        Notifications::set_notification_state(
+            SetNotificationStatePayload {
+                ids: vec![6522],
+                user_id: 1,
+                unread: false,
+            },
+            &mut transaction,
+            &pool,
+        )
+        .await
+        .unwrap();
+
+        // Verify that the object was set to read (eqals seen = true).
+        let uuid = sqlx::query!(r#"SELECT seen FROM notification WHERE id = ?"#, 6522)
+            .fetch_one(&mut transaction)
+            .await
+            .unwrap();
+        assert!(uuid.seen != 0);
+    }
+
+    #[actix_rt::test]
+    async fn set_notification_state_single_id_no_change_if_already_set() {
```

Comment:
```suggestion
    async fn set_notification_state_single_id() {
```

Reply:
set_notification_state_single_id already exists, so basically remove this test?

Reply:
Sorry, I misunderstood your test name then. Maybe `set_notification_state_read` and `set_notification_state_unread` are better names for the two single id cases?

Reply:
What my confusion was: "if_already_set" implies to me that you test what happens when you set an already seen notification to seen. But that's not what the test is doing. Instead, you have two tests, one for the unread case, and one for the seen case (which is good)

Reply:
^^
okay. 

>  "if_already_set" implies to me that you test what happens when you set an already seen notification to seen

That's exactly what I was trying to archive, but maybe I got it wrong?

Notification 6522 is already unread before mutation
I try to set it to unread
and I check if it's still unread after mutation ( `assert!(uuid.seen == 0);` )

Reply:
If it's right, I'll add an comment to explain.

Related: Should I add seperate tests for read/unread?

Reply:
Well, that test doesn't make much sense, then, in my opinion. It would only test if we didn't do some weird toggle logic (which would be inefficient anyways since we needed to select the data previously). So I'd remove that test and rather have a test for unread / read each.

Reply:
check 👍

Reply:
tbh. the single id test does read, the multiple does unread, that's probably enough already?

Reply:
Still would add two tests for that. Single id is the "easy" case where we test the interesting edge cases (so unread/read in this example). And multiple ids is a test that checks if this also works when passing multiple ids (w/o repeating the edge cases again). This makes things easier when stuff breaks (because we'd have more information about what exactly broke).